src/XR/features/WebXRMeshDetector.ts

```diff
@@ -0,0 +1,238 @@
+                const toRemove = this._detectedMeshes.pop();
+                if (toRemove) {
+                    this.onMeshRemovedObservable.notifyObservers(toRemove);
+                }
+            }
+        }
+
+        return true;
+    }
+
+    public dispose(): void {
+        super.dispose();
+        this.onMeshAddedObservable.clear();
+        this.onMeshRemovedObservable.clear();
+        this.onMeshUpdatedObservable.clear();
+    }
+
+    protected _onXRFrame(frame: XRFrame) {
+        if (!this.attached || !frame) {
+            return;
+        }
+
+        const detectedMeshes = frame.detectedMeshes;
+        if (!!detectedMeshes) {
+            const toRemove = this._detectedMeshes
+                .filter((mesh) => !detectedMeshes.has(mesh.xrMesh))
+                .map((mesh) => {
+                    return this._detectedMeshes.indexOf(mesh);
+                });
+            let idxTracker = 0;
+            toRemove.forEach((index) => {
+                const mesh = this._detectedMeshes.splice(index - idxTracker, 1)[0];
+                this.onMeshRemovedObservable.notifyObservers(mesh);
+                idxTracker++;
+            });
+            // now check for new ones
+            detectedMeshes.forEach((xrMesh) => {
+                if (!this._lastDetectedSet.has(xrMesh)) {
+                    const newMesh: Partial<IWebXRMesh> = {
+                        id: meshIdProvider++,
+                        xrMesh: xrMesh,
+                    };
+                    const mesh = this._updateMeshWithXRMesh(xrMesh, newMesh, frame);
+                    this._detectedMeshes.push(mesh);
+                    this.onMeshAddedObservable.notifyObservers(mesh);
+                } else {
+                    // updated?
+                    if (xrMesh.lastChangedTime === this._xrSessionManager.currentTimestamp) {
+                        let index = this._findIndexInMeshArray(xrMesh);
+                        const mesh = this._detectedMeshes[index];
+                        this._updateMeshWithXRMesh(xrMesh, mesh, frame);
+                        this.onMeshUpdatedObservable.notifyObservers(mesh);
+                    }
+                }
+            });
+            this._lastDetectedSet = detectedMeshes;
+        }
+    }
+
+    private _initMeshDetector(session: XRSession) {
+        if (!!session.trySetMeshDetectorEnabled) {
+            session.trySetMeshDetectorEnabled(true);
+        }
+
+        if (!!this._options.preferredDetectorOptions &&
+            !!session.trySetPreferredMeshDetectorOptions) {
+            session.trySetPreferredMeshDetectorOptions(this._options.preferredDetectorOptions);
+        }
+    }
+
+    private _updateMeshWithXRMesh(xrMesh: XRMesh, mesh: Partial<IWebXRMesh>, xrFrame: XRFrame): IWebXRMesh {
+        mesh.xrMesh = xrMesh;
+        if (!this._xrSessionManager.scene.useRightHandedSystem) {
+            mesh.positions = new Float32Array(xrMesh.positions.length);

```

Comment:
why are you creating a new array on each frame? This will be heavy on the garbage collection. If the size changes constantly (which I assume it does), it would be better to use a native js array and keep on expanding it. if it doesn't, there is no need to create a new array. 

Another idea would be to allocate more space than needed space and keep on exponentially increasing it to prevent creating a new Float32Array (which is admittingly faster) while keeping a reference to the actual length. 

Having said all of that, all of those arrays are a direct copy of the arrays in the XRMesh. why not provide them directly? why do we need to recreate them on each frame?

---EDIT---

I just realized this is done because of the right handed system. There is a way to avoid recreating the arrays, which is the way we are loading glTF files (which are right-handed per definition). There is a combination of scaling rotation definition that flips the z correctly, which can be applied to the transformation matrix that the babylon devs will be used (See here - https://github.com/BabylonJS/Babylon.js/blob/master/loaders/src/glTF/2.0/glTFLoader.ts#L481). This is an acceptable solution - adding a root node that the dev won't change, and a child-mesh in it that has this defined geometry.

Reply:
@RaananW do we know whether colliders will be calculated for meshes after applying this right handed to left handed scale transform?

Also, it seems like we may still need to change the indices orders here. Is it better to have a new indices array without updated positions/normals?

Reply:
It's better to have as little as possible changes that will require garbage collection (preferably 0 :-) ). I will leave it up to you, as you know the structure much better than me.

Colliders can be applied, based on the babylon-mesh geometry data, and should be updated when the vertex buffer changes as well. Left-hand vs. right-hand is being checked on each place in the framework where it is needed to differentiate between them, so you need to store according to the scene's handedness, and the other components will take it from there. I hope this answers the question. if it doesn't, please let me know.

Reply:
Is backface culling set as a mesh or material property? IMO, most developers may not know that they will need to enable backface culling to get a valid winding order. I'd prefer to error on the side of exposing the correct winding order when meshes are auto-calculated (now an option when setting things up). If auto-calculations are turned off, folks can go down the backface culling/negative scale path on their own.

---
In reply to: [518369210](https://github.com/BabylonJS/Babylon.js/pull/9307#discussion_r518369210) [](ancestors = 518369210)